src/cpp/server/ServerMain.cpp

```diff
@@ -424,20 +434,52 @@ Error waitForSignals()
       else if (sig == SIGINT || sig == SIGQUIT || sig == SIGTERM)
       {
          LOG_DEBUG_MESSAGE("Received termination signal: " + std::to_string(sig));
-         //
-         // Here is where we can perform server cleanup e.g.
-         // closing pam sessions
-         //
+
+         Error error;
 
          // send SIGTERM signal to specific Workbench child processes
          // this way user processes will not receive the signal until it traverses the tree
-         std::set<std::string> interruptProcs =  {"rsession",
-                                                   "rserver-launcher",
-                                                   "rstudio-launcher",
-                                                   "rworkspaces",
-                                                   "rserver-monitor"};
-
-         Error error = core::system::sendSignalToSpecifiedChildProcesses(interruptProcs, SIGTERM);
+         std::set<std::string> interruptProcs = {
+            "rserver-launcher",
+            "rserver-monitor",
+            "rsession",
+            "rstudio-launcher",
+            "rworkspaces"
+         };
+         
+         // get list of child processes
+         std::vector<system::ProcessInfo> procInfos;
+         error = core::system::getChildProcesses(&procInfos, false);
+         if (error)
+         {
+            LOG_ERROR(error);
+            break;
+         }
+         
+         // pull out pids matching our process names
+         std::vector<pid_t> pids;
+         for (auto&& procInfo : procInfos)
+            if (interruptProcs.count(procInfo.exe))
+               pids.push_back(procInfo.pid);
+         
+         auto waitForChild = [=](int pid)
+         {
+            while (true)
+            {
+               int status = 0;
+               ::waitpid(pid, &status, 0);
+               if (WIFEXITED(status))
+                  break;
+            }
+         };
+         
+         // create threads that will wait for these processes
```

Comment:
I have concerns about starting all these threads, just to make the waitpid calls in parallel. I'm pretty sure the OS maintains those in a queue so one thread ought to be in some ways more efficient. Do you have evidence or something that says we need that?  I do like the idea of having a timeout so a separate thread makes sense for sure.